File: src/main/java/discretemath/bool/expression/compound/ExpressionTreeNode.java
package discretemath.bool.expression.compound;

import discretemath.bool.expression.atomic.AtomicBooleanExpression;
import discretemath.bool.expression.BooleanExpression;

public sealed interface ExpressionTreeNode permits OperatorNode, ValueNode {

	int getHeight();

	static ExpressionTreeNode fromBooleanExpression(BooleanExpression expression) {
		if (expression instanceof AtomicBooleanExpression atomicBooleanExpression) {
			return new ValueNode(atomicBooleanExpression);
		} else if (expression instanceof CompoundBooleanExpression compoundBooleanExpression) {
			return compoundBooleanExpression.getRootNode();
		} else {
			throw new UnsupportedOperationException(expression.getClass().toString());
		}
	}

	static ExpressionTreeNode[] fromBooleanExpression(BooleanExpression[] expressions) {
		ExpressionTreeNode[] nodes = new ExpressionTreeNode[expressions.length];
		for (int i = 0; i < expressions.length; i++) {
			BooleanExpression expression = expressions[i];
			nodes[i] = fromBooleanExpression(expression);
		}

		return nodes;
	}

	static BooleanExpression[] toBooleanExpression(ExpressionTreeNode[] nodes) {
		BooleanExpression[] operandExpressions = new BooleanExpression[nodes.length];
		for (int i = 0; i < nodes.length; i++) {
			ExpressionTreeNode operandNode = nodes[i];

			operandExpressions[i] = toBooleanExpression(operandNode);
		}

		return operandExpressions;
	}

	static BooleanExpression toBooleanExpression(ExpressionTreeNode node) {
		if (node instanceof ValueNode valueNode) {
			return valueNode.expression();
		} else {
			return new CompoundBooleanExpression((OperatorNode) node);
		}
	}
}

File: src/main/scala/discretemath/bool/expression/compound/ExpressionTreeNode.scala
package discretemath.bool.expression.compound

import discretemath.bool.expression.BooleanExpression
import discretemath.bool.operator.Operator
import discretemath.bool.operator.Operators.{BICONDITIONAL, CONDITIONAL, NOT}

sealed trait ExpressionTreeNode {
  def getHeight: Int
}

object ExpressionTreeNode {
  def fromBooleanExpression(expression: BooleanExpression): ExpressionTreeNode = {
    expression match {
      case compoundExpression: CompoundBooleanExpression => compoundExpression.rootNode
      case _ => ValueNode(expression)
    }
  }

  def fromBooleanExpression(expressions: collection.Seq[BooleanExpression]): Seq[ExpressionTreeNode] = expressions.map(fromBooleanExpression).toSeq

  def toBooleanExpression(node: ExpressionTreeNode): BooleanExpression = {
    node match {
      case ValueNode(expression) => expression
      case operatorNode: OperatorNode => new CompoundBooleanExpression(operatorNode)
    }
  }

  def toBooleanExpression(nodes: collection.Seq[ExpressionTreeNode]): Seq[BooleanExpression] = nodes.map(toBooleanExpression).toSeq
}

case class OperatorNode(operator: Operator, operands: Seq[ExpressionTreeNode]) extends ExpressionTreeNode {

  override def getHeight: Int = 1 + operands.map(o => o.getHeight).max

  override def toString: String = operator.toString(ExpressionTreeNode.toBooleanExpression(operands))
}

object OperatorNode {

  def *(operator: Operator, operands: ExpressionTreeNode*): OperatorNode = OperatorNode(operator, operands)

  def not(operand: ExpressionTreeNode): OperatorNode = new OperatorNode(NOT, Seq(operand))

  def conditional(hypothesis: ExpressionTreeNode, conclusion: ExpressionTreeNode): OperatorNode = new OperatorNode(CONDITIONAL, Seq(hypothesis, conclusion))

  def biconditional(operand1: ExpressionTreeNode, operand2: ExpressionTreeNode): OperatorNode = new OperatorNode(BICONDITIONAL, Seq(operand1, operand2))
}

case class ValueNode(expression: BooleanExpression) extends ExpressionTreeNode {
  override def getHeight = 0

  override def toString: String = expression.toString
}
